packages/headless/src/features/search/search-actions.ts

```diff
@@ -160,6 +160,18 @@ export const executeSearch = createAsyncThunk<
       return rejectWithValue(retried.response.error);
     }
 
+    const retriedResponse = retried.response.success;
+    analyticsAction(
+      dispatch,
+      () =>
+        getStateAfterResponse(
+          correctedQuery,
+          retried.duration,
+          state,
+          retriedResponse
+        ),
+      extra
+    );
```

Comment:
This is where the magic happens.

I'm a redux noob, so it took me way too long to figure out, but I can mock the future state that the action expects this way. It's a bit of a dirty fix, but if it needs to be replaced in the future, that should be easy enough.

Reply:
It is a bit of a dirty fix indeed, but IMHO it's way better than the other thing you had before ;) With the array of analytics action.

I think the real problem lies in the AnalyticsProvider class, which extracts (`pulls`) from the redux state directly, which seemed like a good idea at the beginning.

I think it would be easier to change it to a `push` mechanism, where the consumer of the analytics provider instead map from some object (the state, a search API response, etc) to a predefined interface.

It would also solve another problem we had to dirty fix for recommendations.

Reply:
I'd strongly encourage looking into what it would take to pass data in.

If we think of the big picture, the metadata for an analytics call comes from one of two places. Either,

1. The existing state of the page before the request.
2. The new data we get from response.

I think a more flexible analytics api would allow us to pass the data in explicitly. This would make the api independent of the order of operations.

Versus the current approach where we have to dispatch the analytics action at the precise moment, once the state is holding the correct combination of before & after. The `getStateAfterResponse` function is essentially bending time, returning the new state before it's been applied. It's a good work-around, but one we could avoid if we could pass data in explicitly.